### ISSUE TITLE
UPnP NAT Rules Not Working

### ISSUE THREAD

**Issue by effex12** [opnsense/plugins#3241]
When using UPnP with 2 PCs on the same LAN playing Destiny 2, UPnP will correctly set up the port mappings but not create NAT rules for routing the traffic. 

Destiny 2 will choose UDP port 3097 for the first PC and UDP port 3098 for the second PC. These reservations can be seen in the Services/Universal Plug and Play/Status section. 

To note, these PCs are set up with static DHCP leases. 

If static port mapping is set up on the subnet as an outbound NAT rule (192.168.1.0/24), the game will show OPEN NAT. However, something behind the scenes with miniupnpd is not working as traffic does not get routed properly.

If static port mapping is DISABLED, you will get STRICT NAT regardless of the port mapping in UPnP. 

I believe this issue was uncovered on the netgate forum and a fix was provided by the miniupnpd developer which has been added as a patch for future versions of pfsense. 

I'm not sure exactly how opnsense differs from pfsense under the hood, but the solution appears to involve adding rules before the outbound NAT rules that point to miniupnp and setting up a binat anchor. 

See here:

![image](https://user-images.githubusercontent.com/29265285/210087273-01c8b67b-d23c-4349-9e74-fe6f892d105c.png)

I'm happy to provide logs or any other information that could be helpful but I've done a lot of research and I'm convinced there is a bug inherent in the miniupnpd and freebsd implementation. 

Please reference these links here as I believe they are describing the same bug that I am experiencing:

https://forum.netgate.com/topic/169837/upnp-fix-for-multiple-clients-consoles-playing-the-same-game

https://redmine.pfsense.org/issues/7727

**Comment by fichtner** [comment]
Allegedly the dev version has all the changes required.

**Comment by fichtner** [comment]
From the console:

    # pkg install os-upnp-devel

**Comment by effex12** [comment]
> From the console:
> 
> ```
> # pkg install os-upnp-devel
> ```

Thank you so much for your help. I will try this! I assume I'll also need to remove the existing package? 

I can test later and validate that the ports map correctly and static port mapping isn't needed.

Thank you!

**Comment by fichtner** [comment]
Existing will be replaced. Configuration is the same though.

**Comment by effex12** [comment]
I can confirm that this fixes the issues and no static port NAT rules are required. 

This fix allows you to achieve an OPEN NAT with multiple PCs without any specific rules other than UPnP enabled for the specific /32 ip address. 

This should probably be stickied somewhere as a solution as there is a lot of misinformation around what is required to make UPnP work when port forwarding is NOT an option. 

Thank you so much @fichtner!

**Comment by effex12** [comment]
I guess I spoke too soon. 

It looks like this fix will open the right ports and forward them correctly but it interferes with opnsense's rule ordering. 

Now, a lot of legitimate traffic is showing up blocked by the default deny / state violation rule that is a floating rule. 

For instance, I'm seeing blocks from 192.168.1.x to legitimate destinations on TCP 443. I would think my allow LAN to all default rule would pass that traffic but it's not. 

 I would assume it's an issue with the NAT rules and their priority while being evaluated. 

Thanks!

**Comment by fichtner** [comment]
According to @Tawmu this seemed to be correct, but perhaps an overlap is still occurring. Might be misconception about rules required or wrong hook sequence in edge case.

**Comment by Tawmu** [comment]
I'm surprised - we ran these fixes at a recent LAN party with over 700 devices and didn't see the issues mentioned here. I'll see if I can replicate.

**Comment by effex12** [comment]
> I'm surprised - we ran these fixes at a recent LAN party with over 700 devices and didn't see the issues mentioned here. I'll see if I can replicate.
> 
> @effex12 could you show us your firewall/NAT rules or send a (sanitised) config backup at all?

I'm working on sanitizing the config file, without going line by line, not sure how to verify I'm not sending anything out that would compromise my accounts/ip etc. 

To note, even before this fix was available, I could get the game to say "OPEN NAT" by using UPnP and static port mapping for the two PCs. Even though it said that in the UI, there were still issues in the same game (I have videos of on my pc the other player running into walls, other random problems that don't exist outside of being in the SAME match)

When I first applied the fix, I disabled the outbound NAT rules that I created for static port mapping on the 192.168.1.0/24 network. 

That allows me to get OPEN NAT in the game on both PCs which wasn't possible before, however after validating the config where we were both in the same game, the lag issues persisted. (They don't occur when we aren't in the same match. It's not my connection or latency issue as I've tested all that. I'm hardwired gigabit fiber on att and there are no issues when we don't play together.)

Enabling static port mapping for the subnet 192.168.1.0/24 also did not work. 

(I always cleared states on the firewall when I made changes) 

From there, I tried going to Firewall/Advanced/NAT and enabling Reflection for port forwards, Reflection for 1:1 and Automatic outbound NAT for Reflection as I thought maybe it's some sort of issue with routing the traffic back into the network. 

At that point, I received a "Bad Packet" error from Battle Eye (Anti Cheat), so I disabled all of the above and bounced the PCs and the firewall. 

This is my outbound NAT config as of now:

![image](https://user-images.githubusercontent.com/29265285/210151535-4cd5f3c0-438e-472e-9f72-a117b943eece.png)


No special firewall rules:

![image](https://user-images.githubusercontent.com/29265285/210151575-2f45c71d-7b24-4edd-b16a-adb6fa6ddb9e.png)

**Comment by effex12** [comment]
Happy New Year all!

Just to provide an update, I plugged in the ER605 (I believe it's linux based) and enabled UPnP there. I performed the same tests that I did using opnsense and we didn't experience any of the problems being in the same match. I suspect there is a routing issue with the NAT rules and miniupnpd with freebsd but I'm not an expert on this. 

I'm not sure if the fix is similar to what I linked in the netgate forum, but I noticed their code creates a "binat-anchor" and the code fix here doesn't appear to do that. 

Thanks again for the help. As always, I'm willing to test alternate configurations in support of finding a permanent solution as I do really enjoy opnsense. 

This may be an edge case that has to do with Destiny 2 and it's implementation of UPnP but according to their docs, UPnP is the only way to use multiple clients on the same LAN.

**Comment by fichtner** [comment]
Could be an additional anchor missing as comparing diffs is inconclusive between different code cases, though binat might not be what you’re looking for here.

**Comment by effex12** [comment]
Update:

I’m running into the same problems with the ER605. 

I can’t pinpoint exactly what’s happening but I believe internally Destiny 2 is using UDP 3097 for a single PC and then grabs 3098 with UPnP.

I believe externally this gets mapped to my external IP and port 3074 but I’m not 100% sure on that part.

I tried doing packet capture and looking in wireshark and I can’t find any smoking gun.

I found an old post where someone mentioned it could have something to do with UDP session timeout. 

Does anyone have any ideas on other things to try? 

I’m at a complete loss as to how to correct this issue.

Other forums have talked about full cone NAT being necessary but I’m not really sure if that’s true and if so what I would need to do with my configuration to get this to work. 

To note, the game is showing OPEN NAT. But UDP packets are being lost between the two PCs. 

Thank you!

**Comment by Tawmu** [comment]
@effex12 I've not had a chance to play around with your issue yet but port 3074 is used by Xbox Live as far as I know. We've had instances in the past with XBL games where we've had to block 3074 mapping using the UPNP ACL option to force clients to request other ports.

**Comment by effex12** [comment]
@Tawmu no worries. 

If UPnP and the NAT is problematic with Destiny 2 on multiple implementations, could I just get a block of static IPs from AT&T and use port forwarding with each PC NAT’d to a different public IP? Would that actually solve it?

**Comment by effex12** [comment]
Just to provide an update --

I believe the problem is hairpin NAT with the mini upnp implementation. 

When two pcs are trying to play Destiny 2 using UPNP and ports 3097 UDP and 3098 UDP, there are packets lost between the two. 

I confirmed this behavior also existed with the ER605 which is linux based. 

I ordered a static ip block from ATT and put each PC on a different Virtual IP with 1:1 NAT and there are no problems. 

I believe there is some sort of issue with upnp and NAT loopback/hairpin.

**Comment by effex12** [comment]
Just as an update, after all the troubleshooting that was done here, I found out about a month later that the problem was on Steam's side:

**_"Valve has constantly been making improvements to its Steam Datagram Relay (SDR) service to provide safe, high-quality Destiny 2 connections. In this case, they found new relays which were not configured properly and severely impacted network quality between players during peak hours. The problematic relays were taken offline at 3 PM last Friday pending a more thorough investigation, and this appears to have resolved the immediate issue for players."_**

I believe the proposed fix solved the UPNP issue. However, since updating to the latest version of opnsense, the port reservations aren't being made. I haven't confirmed that it's an issue with opnsense yet. It could be a problem on the game's side. I will open a separate issue once I validate it's not an issue with the game. 

Thanks!
